### PR TITLE
Feature/react auth0

### LINKED ISSUE #507: Logout and Cancel button on Settings (almost) not clickable
**Describe the bug**
If the website is viewed on smaller screen, some buttons of the modal settings dialog are not accessible to click.

**Is it on a Production or Development Build?**
develop

**To Reproduce**
1. Login
2. Go to "Me"
3. Open Settings, try to click "Cancel" or "Logout"

**Expected behavior**
Allow full button area to be clickable

**What actually happens**
Only parts or nothing of the button area is clickable.

**Screenshots**
Marked the clickable area with red rectangle:
Bildschirmfoto 2020-11-28 um 17 37 10
Bildschirmfoto 2020-11-28 um 17 35 55

**Desktop (please complete the following information):**
- OS: latest MacOS
- Browser: latest Chrome

### PR DESCRIPTION
**Pull request by harshvitra**
Fixes #507

Replaces Auth0 Setup

TODO:
- [x] On Vercel still login and logout are a problem (maybe just because of the dynamic Vercel URLs) https://github.com/Plant-for-the-Planet-org/planet-webapp/pull/498#issuecomment-734824586 and https://github.com/Plant-for-the-Planet-org/planet-webapp/pull/498#issuecomment-735236715

### CODE REVIEW

**Comment by sagararyal** [comment]
Few bugs:

Logging in from a Url that is not VERCEL URL returns Invalid State in the header

Screen Shot 2020-11-27 at 9 32 18 AM

Reloading a page eg: manage projects when logged in, redirects the user to homepage.
https://planet-webapp-lglgt2i0t.vercel.app/manage-projects/proj_WZkyugryh35sMmZMmXCwq7YY

File: src/features/common/TreeCounter/TreeCounter.tsx [hunk 1]
```diff
@@ -83,7 +83,8 @@ export default function TpoProfile(props: any) {
           <h1>{Sugar.Number.abbr(Number(props.planted), 1)}</h1>
           <h2>{t('me:treesPlanted')}</h2>
         </div>
-        {props.target && props.target !== 0 && (
+        
+        {props.target ? props.target !== 0 && (
```
Comment (norbertschuler) [inline src/features/common/TreeCounter/TreeCounter.tsx:87]: Where is the difference between `props.target !== 0` && and `props.target` - I wonder if the part `props.target !== 0 &&` might be useless?

File: src/features/donations/components/treeDonation/PaymentFunctions.ts [hunk 2]
```diff
@@ -12,10 +12,10 @@ export async function createDonation(data: any,session:any) {
         : 'en'
     }`
   }
-  if(session){
+  if(token && token !== ''){
```
Comment (norbertschuler) [inline src/features/donations/components/treeDonation/PaymentFunctions.ts:15]: same as above - do we need to check `&& token !== ''`?

**Comment by norbertschuler** [comment]
- [ ] If you reload a page, e.g http://localhost:3000/t/norberts-baumschule or http://localhost:3000/manage-projects/add-project or http://localhost:3000/manage-projects/proj_xDwYRB6yGOWiuN4DupGO67ZA in the browser your login status seems to be forgotten and the app handles this urls like if you are not logged in:
Bildschirmfoto 2020-11-27 um 13 05 25

**Comment by norbertschuler** [comment]
On https://planet-webapp-git-feature-reactauth0.planetapp.vercel.app/ I cannot test signup (it only offers login) and on localhost the link in the verification mail from auth0 redirects to https://localhost/ (while I was using http://localhost:3000/), so I ended up in an invalid state for my signup.

(1) after trying to signup before verifying email:
Bildschirmfoto 2020-11-27 um 13 50 15

(2) after trying to verify the email:
Bildschirmfoto 2020-11-27 um 13 54 29
I will to try to setup a local SSL proxy again.

File: src/features/donations/components/treeDonation/PaymentFunctions.ts (hunk 2 above)
Comment (harshvitra) [reply src/features/donations/components/treeDonation/PaymentFunctions.ts:15]: Sometimes '' is not considered as Boolean false.

File: src/features/common/Layout/Navbar/index.tsx [hunk 3]
```diff
@@ -15,72 +14,86 @@ import { ThemeContext } from '../../../../theme/themeContext';
...
+  // This effect is used to get and update UserInfo if the isAuthenticated changes
+  React.useEffect(() => {
+    async function loadFunction() {
+      const token = await getAccessTokenSilently();
+      setToken(token);
+      let userInfo;
+      userInfo = await getUserInfo(token, router, logout);
+      setUserInfo(userInfo);
+    }
+    if (isAuthenticated) {
```
Comment (norbertschuler) [inline src/features/common/Layout/Navbar/index.tsx:48]: If you login and afterwards open https://localhost/ again in your browser `isAuthenticated` is `false` and `isLoading` is still `true` and this code is never called again! So if you try to open the profile afterwards, you get presented the login again. 

(If you login for the first time, this code is executed twice. On the first run also with `isAuthenticated: false` and `isLoading: true`, but on the second run with  `isAuthenticated: true` and `isLoading: false`, so the user info gets loaded.)

Comment (norbertschuler) [reply src/features/common/Layout/Navbar/index.tsx:48]: changed auth0 cacheLocation to localstorage

**Comment by norbertschuler** [comment]
> * [ ]  On https://planet-webapp-git-feature-reactauth0.planetapp.vercel.app/ I cannot test signup (it only offers login) and on localhost the link in the verification mail from auth0 redirects to https://localhost/ (while I was using http://localhost:3000/), so I ended up in an invalid state for my signup.

This is the error I got in the NavBar component from Auth0 (testing locally):

Bildschirmfoto 2020-11-28 um 15 16 12

I could fix the error with a patch https://github.com/Plant-for-the-Planet-org/planet-webapp/compare/feature/reactAuth0...feature/reactAuth0SPA - but this does not work in Vercel (Login with Auth0). Maybe there is another solution?

**Comment by norbertschuler** [comment]
- [ ] Logout showing me this URL https://accounts.plant-for-the-planet.org/v2/logout?client_id=8EyITuZvf63r8J8mjuYoaJ5Z1GJVm3zc&auth0Client=eyJuYW1lIjoiYXV0aDAtcmVhY3QiLCJ2ZXJzaW9uIjoiMS4yLjAifQ%3D%3D with result `OK` instead of redirecting to logged out website.

**Comment by norbertschuler** [comment]
- [ ] Educational profiles (only Individual and TPO are currently supported) show an empty profile page.

**Comment by norbertschuler** [comment]
- [ ] https://localhost/t/norbert-schuler and other treecounters can only be shown for logged in users as otherwise a 401 error is returned by the app/profile call, right?

**Comment by harshvitra** [comment]
> Reloading a page eg: manage projects when logged in, redirects the user to homepage.

This issue is not happening please check via latest deployments

**Comment by harshvitra** [comment]
> Logging in from a Url that is not VERCEL URL returns Invalid State in the header

Can you tell on which URL you are testing

**Comment by sarveshpro** [comment]
app donation URLs - `/donation-payment/OEJQUS1YFF` not working
image

File: src/features/user/UserProfile/components/SettingsModal.tsx [hunk 4]
```diff
@@ -30,7 +30,9 @@ export default function SettingsModal({
 
   const logoutUser = () => {
     localStorage.removeItem('userInfo');
-    logout();
+    if (typeof window !== 'undefined') {
+      logout({returnTo:window.location.href});
+    }
```
Comment (norbertschuler) [inline src/features/user/UserProfile/components/SettingsModal.tsx]: There a more place in our code doing this kind of logout() - @harshvitra do we need to adapt them all like this?

Comment (harshvitra) [reply src/features/user/UserProfile/components/SettingsModal.tsx]: Only if it works, right now it says that we need to add this Log out URL inside Auth0 dashboard

**Comment by norbertschuler** [comment]
> * [x]  https://planet-webapp-git-feature-reactauth0.planetapp.vercel.app/t/norbert-schuler and other treecounters can only be shown for logged in users as otherwise a 401 error is returned by the `app/profile` call (the slug `norbert-schuler` seems to get lost) - before the treecounters also could be looked at without being logged in!

- [ ] As an not logged in user I am getting an error loading https://planet-webapp-git-feature-reactauth0.planetapp.vercel.app/t/norbert-schuler as following:

```
pages/t/[id].tsx (87:39) @ _callee2$

  85 | let userInfo;
  86 | userInfo = await getUserInfo()
> 87 | let currentUserSlug = userInfo.slug ? userInfo.slug : null;
     |                               ^
  88 | 
  89 | // some user logged in and slug matches -> private profile
  90 | if (!isLoading && token && currentUserSlug === slug) { 
```

**Comment by norbertschuler** [comment]
> app donation URLs - `/donation-payment/OEJQUS1YFF` not working

- [ ] Same as https://planet-webapp-git-feature-reactauth0.planetapp.vercel.app/donation-payment/OEJQUS1YFF also happens for https://planet-webapp-git-feature-reactauth0.planetapp.vercel.app/all if called directly from the link. Using the new <Auth0Provider>-component is messing up the layout if opening links directly, but working if changing to another path with e.g. the menu items of the Navbar.

**Comment by norbertschuler** [comment]
> > app donation URLs - `/donation-payment/OEJQUS1YFF` not working
> 
> * [ ]  Same as https://planet-webapp-git-feature-reactauth0.planetapp.vercel.app/donation-payment/OEJQUS1YFF also happens for https://planet-webapp-git-feature-reactauth0.planetapp.vercel.app/all if called directly from the link. Using the new `<Auth0Provider>`-component is messing up the layout if opening links directly, but working if changing to another path with e.g. the menu items of the Navbar.

I was able to stop the _repetition of the former div id and styles into the next div if using a fragment placeholder_ (that was the bug!) by swapping the `<>` with a custom tag `<planetapp>`, but this solutions feeds weird and maybe has to be applied to all other usages of `<>` in other pages which do not work if loaded directly as well.

**Comment by norbertschuler** [comment]
- [ ] Build fails as new pages https://github.com/Plant-for-the-Planet-org/planet-webapp/blob/62f9d2c9dd8fe7d6eeda1b5a6e31103dd010b242/pages/c/%5Bid%5D.tsx#L1 and https://github.com/Plant-for-the-Planet-org/planet-webapp/blob/62f9d2c9dd8fe7d6eeda1b5a6e31103dd010b242/pages/register-trees.tsx#L1 still need to be changed to use @auth0/auth0-react.

**Comment by norbertschuler** [comment]
@harshvitra can you please try to merge develop. I am not sure about how to correctly solve the conflicts.

**Comment by harshvitra** [comment]
> @harshvitra can you please try to merge develop. I am not sure about how to correctly solve the conflicts.

On it

**Pull request merged**